src/dead_letters.rs

```diff
@@ -0,0 +1,173 @@
+            _ => unreachable!(),
+        }
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum DeadLetterQueueError {
+    #[error("JSON serialization failed: {source}")]
+    SerdeJson {
+        #[from]
+        source: serde_json::Error,
+    },
+
+    #[error("Delta write failed: {source}")]
+    DeltaWriter {
+        #[from]
+        source: DeltaWriterError,
+    },
+}
+
+pub struct DeadLetterQueueOptions {
+    /// Table URI of the delta table to write dead letters to. Implies usage of the DeltaSinkDeadLetterQueue.
+    pub delta_table_uri: Option<String>,
+}
+
+#[async_trait]
+pub trait DeadLetterQueue: Send + Sync {
+    async fn write_dead_letter(
+        &mut self,
+        dead_letter: DeadLetter,
+    ) -> Result<(), DeadLetterQueueError> {
+        self.write_dead_letters(vec![dead_letter]).await
+    }
+
+    async fn write_dead_letters(
+        &mut self,
+        dead_letters: Vec<DeadLetter>,
+    ) -> Result<(), DeadLetterQueueError>;
+}
+
+pub async fn dlq_from_opts(
+    options: DeadLetterQueueOptions,
+) -> Result<Box<dyn DeadLetterQueue>, DeadLetterQueueError> {
+    if let Some(table_uri) = options.delta_table_uri {
+        Ok(Box::new(
+            DeltaSinkDeadLetterQueue::for_table_uri(table_uri.as_str()).await?,
+        ))
+    } else {
+        Ok(Box::new(NoopDeadLetterQueue {}))
+    }
+}
+
+pub struct NoopDeadLetterQueue {}
+
+#[async_trait]
+impl DeadLetterQueue for NoopDeadLetterQueue {
+    async fn write_dead_letters(
+        &mut self,
+        _dead_letters: Vec<DeadLetter>,
+    ) -> Result<(), DeadLetterQueueError> {
+        // noop
+        Ok(())
+    }
+}
+
+pub struct LoggingDeadLetterQueue {}
+
+#[async_trait]
+impl DeadLetterQueue for LoggingDeadLetterQueue {
+    async fn write_dead_letters(
+        &mut self,
+        dead_letters: Vec<DeadLetter>,
+    ) -> Result<(), DeadLetterQueueError> {
+        for dead_letter in dead_letters {
+            info!("DeadLetter: {:?}", dead_letter);
+        }
+
+        Ok(())
+    }
+}
+
+pub struct DeltaSinkDeadLetterQueue {
+    delta_writer: DeltaWriter,
+}
+
+impl DeltaSinkDeadLetterQueue {
+    pub async fn for_table_uri(table_uri: &str) -> Result<Self, DeadLetterQueueError> {
+        Ok(Self {
+            delta_writer: DeltaWriter::for_table_path(table_uri).await?,
+        })
+    }
+}
+
+#[async_trait]
+impl DeadLetterQueue for DeltaSinkDeadLetterQueue {
+    async fn write_dead_letters(
+        &mut self,
+        dead_letters: Vec<DeadLetter>,
+    ) -> Result<(), DeadLetterQueueError> {
+        let values: Result<Vec<Value>, _> = dead_letters
+            .iter()
+            .map(|dl| serde_json::to_value(dl))
+            .collect();
+        let values = values?;
+
+        info!("Starting insert_all");
+        let version = self.delta_writer.insert_all(values).await?;
+
+        // TODO: take opt for checkpoint creation
+        if version % 10 == 0 {
+            // TODO: create checkpoint on every 10th version
```

Comment:
I presume this TODO is going to be addressed inside of this change?


But also, doesn't the delta writer handle that for us already?

Reply:
TODO been addressed. The DeltaWriter does not handle checkpoints automatically. We provide an option and invoke it explicitly if set.